app/javascript/utils/react-bootloader.jsx

```diff
@@ -53,55 +91,62 @@ const render = (elem, component) => {
     </React.StrictMode>,
     elem,
     () => {
-      setTimeout(() => {
-        elem.classList.add('--hydrated')
-      }, 1)
+      // console.log(Date.now(), 'rendered')
+      elem.classList.add('--hydrated')
     }
   )
 
   const unloadOnce = () => {
     ReactDOM.unmountComponentAtNode(elem)
     document.removeEventListener('turbo:before-render', unloadOnce)
   }
-  document.addEventListener('turbo:before-render', unloadOnce)
+  // document.addEventListener('turbo:before-render', unloadOnce)
 }
 
-const renderComponents = (parentElement, mappings) => {
+function renderComponents(parentElement, mappings) {
+  // console.log(Date.now(), 'renderComponents()')
   if (!parentElement) {
     parentElement = document.body
   }
 
-  for (const [name, generator] of Object.entries(mappings)) {
-    const selector = '[data-react-' + name + ']'
-    parentElement.querySelectorAll(selector).forEach((elem) => {
-      const data = JSON.parse(elem.dataset.reactData)
-      render(elem, generator(data, elem))
-    })
+  const elems = Array.from(
+    parentElement.getElementsByClassName('c-react-component')
+  )
+  for (let elem of elems) {
+    const reactId = elem.dataset['reactId']
+    const generator = mappings[reactId]
+    if (!generator) {
+      continue
+    }
+
+    const data = JSON.parse(elem.dataset.reactData)
+    render(elem, generator(data, elem))
   }
 }
 
-const renderTooltips = (parentElement, mappings) => {
+function renderTooltip(mappings, elem) {
+  const name = elem.dataset['tooltipType'] + '-tooltip'
+  const generator = mappings[name]
+
+  if (!generator) {
+    return
+  }
+  const component = generator(elem.dataset, elem)
+
+  const tooltipElem = document.createElement('span')
+  elem.insertAdjacentElement('afterend', tooltipElem)
+
+  render(tooltipElem, <ExercismTippy content={component} reference={elem} />)
+}
+
+function renderTooltips(parentElement, mappings) {
   if (!parentElement) {
     parentElement = document.body
   }
 
   parentElement
     .querySelectorAll('[data-tooltip-type][data-endpoint]')
     .forEach((elem) => {
-      const name = elem.dataset['tooltipType'] + '-tooltip'
-      const generator = mappings[name]
-
-      if (!generator) {
-        return
-      }
-      const component = generator(elem.dataset, elem)
-
-      const tooltipElem = document.createElement('span')
-      elem.insertAdjacentElement('afterend', tooltipElem)
-
-      render(
-        tooltipElem,
-        <ExercismTippy content={component} reference={elem} />
-      )
+      renderTooltip(mappings, elem)
```

Comment:
Instead of 

```js
.forEach((elem) => {
  renderTooltip(mappings, elem)
})
```

you can do:

```js
.forEach((elem) => renderTooltip(mappings, elem))
```

Reply:
I thought our linter would do this automatically for me 😢